pkg/csr/csr.go

```diff
@@ -0,0 +1,248 @@
+	var hasMatch bool
+	for _, address := range c.instanceAddresses {
+		if c.areAddressesEqual(address, nodeName) {
+			hasMatch = true
+			break
+		}
+	}
+	if !hasMatch {
+		// CSR is not from a BYOH Windows instance, don't deny it or return error
+		// since it might be from a linux node, instead log if returns false here
+		return false, nil
+	}
+	// Kubelet on a node needs two certificates for its normal operation:
+	// Client certificate for securely communicating with the Kubernetes API server
+	// Server certificate for use by Kubernetes API server to talk back to kubelet
+	// Both types are validated based on their contents
+	var validated bool
+	if c.isNodeClientCert(parsedCSR) {
+		// Node client bootstrapper CSR is received before the instance becomes a node
+		// hence we should not proceed if a corresponding node already exists
+		_, err := c.k8sClientSet.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "unable to get node %s", nodeName)
+		} else if err == nil {
+			return false, errors.Wrapf(err, "%s node already exists, cannot validate CSR", nodeName)
+		}
+	} else {
+		if validated, err = c.validateKubeletServingCSR(parsedCSR); !validated {
+			return false, errors.Wrapf(err, "unable to validate kubelet serving CSR %s", c.Name)
+		}
+	}
+	return true, nil
+}
+
+// validateKubeletServingCSR validates a kubelet serving CSR for its contents
+func (c *Approver) validateKubeletServingCSR(csr *x509.CertificateRequest) (bool, error) {
+	if c.CertificateSigningRequest == nil || csr == nil {
+		return false, errors.New("CSR or request should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(c.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR %s contains invalid number of groups: %d", c.Name, len(c.Spec.Groups))
+	}
+	groups := sets.NewString(c.Spec.Groups...)
+	if !groups.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR %s does not contain required groups", c.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !c.hasUsages(kubeletServerUsages) {
+		return false, errors.Errorf("CSR %s does not contain required usages", c.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR %s does not contain required subject organization", c.Name)
+	}
+	return true, nil
+}
+
+// Approve approves a CSR by updating it's status conditions to true if it is a valid CSR
+func (c *Approver) Approve() error {
+	if validated, err := c.validateCSRContents(); !validated {
+		c.log.Info("CSR contents are invalid for approval by WMCO CSR Approver", "CSR", c.Name)
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", c.Name)
```

Comment:
Will we constantly keep trying to reconcile this invalid CSR?

Reply:
no we wouldn't, this will return error only if the err in `return errors.Wrapf(err, "could not validate CSR %s contents for approval", c.Name)` is not nil, which could be returned due to parsing errors. If CSR is not a BYOH CSR we are only logging here.